### ISSUE TITLE
[yul optimizer]: Dead code eliminator may produce incorrect AST because it cannot remove references to dead variables in blocks other than the one containing the terminating opcode

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#6492]
## Description

For

```
{
  for { stop() let i_0 := 0 } lt(i_0,2) { i_0 := add(i_0,1) }
  {
    let i_1 := i_0
  }
}
```

The dead code eliminator optimizer removes the variable declaration following the `stop()` opcode in the for-loop initializer block, but does not remove references to this variable (and consequently the statements that depend on this reference) in the for-loop-expression and body blocks.

## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ cd build && make yultopi
$ ./test/tools/yultopi --input-file <code.yul>
(q)quit/(f)flatten/(c)se/initialize var(d)ecls/(x)plit/(j)oin/(g)rouper/(h)oister/
  (e)xpr inline/(i)nline/(s)implify/varname c(l)eaner/(u)nusedprune/ss(a) transform/
  (r)edundant assign elim./re(m)aterializer/f(o)r-loop-pre-rewriter/
  s(t)ructural simplifier/equi(v)alent function combiner/ssa re(V)erser/? 
  stack com(p)ressor/(D)ead code eliminator/? 
D
 D
----------------------
{
    for {
        stop()
    }
    lt(i_0, 2)
    {
        i_0 := add(i_0, 1)
    }
    {
        let i_1 := i_0
    }
}
Error analyzing source.
:5:8: Error: Identifier not found.
    lt(i_0, 2)
       ^-^
:10:20: Error: Identifier not found.
        let i_1 := i_0
                   ^-^
:7:20: Error: Identifier not found.
        i_0 := add(i_0, 1)
                   ^-^
:7:9: Error: Variable not found or variable not lvalue.
        i_0 := add(i_0, 1)
        ^-^
```

Notes:
  - Funny that `solc` itself does not reproduce this issue.
  - However, I could make `solc` trigger this issue with a nested for loop like so

```
{
  {
    for { let i_0 } lt(0,1) { i_0 := add(i_0,0) }
    {
      for {stop() let i_1 } lt(1,0) { i_0 := add(i_1,0) }
      {}
    }
  }
}
```

Then, solc throws an OptimizerException from here

https://github.com/ethereum/solidity/blob/ef3a18999c840c523cce73f6074e816334b5c753/libyul/optimiser/CommonSubexpressionEliminator.cpp#L65

I asked myself why `solc` produces correct code for the original example, but I couldn't find out why. My guess is that the dead code eliminator bug is hidden (masked rather) by some other optimization step(s).

**Comment by sifmelcara** [comment]
The bug can be fixed by executing `ForLoopInitRewriter` before `DeadCodeEliminator`. So we don't need to worry about the scoping rule exception for "for loop init".

> I asked myself why solc produces correct code for the original example, but I couldn't find out why.

I just traced some optimizer code and found the reason is that
https://github.com/ethereum/solidity/blob/ef3a18999c840c523cce73f6074e816334b5c753/libyul/optimiser/Suite.cpp#L71-L114

although they both produce incorrect AST after line 74's `DeadCodeEliminator`, only the first example managed to not throw any error until line 112's `DeadCodeEliminator`. `ForLoopInitRewriter` is executed at line 78, so line 122's `DeadCodeEliminator` is able to remove those statements which containing undefined variable.

Later I will open a PR to fix this.

**Comment by bshastry** [comment]
Thank you @sifmelcara for helping with this!

**Comment by bshastry** [comment]
@sifmelcara Can this issue be closed? IIUC, issues #6480 and #6495 are not dependent on this issue that PR #6494 fixed.

**Comment by sifmelcara** [comment]
Yea I think so :)
